ash-swapchain/src/lib.rs

```diff
@@ -0,0 +1,388 @@
+                                None,
+                            )
+                            .unwrap(),
+                        acquire: device
+                            .create_semaphore(&vk::SemaphoreCreateInfo::default(), None)
+                            .unwrap(),
+                        generation: 0,
+                    }
+                })
+                .collect(),
+            frame_index: 0,
+
+            surface,
+            physical_device,
+            handle: vk::SwapchainKHR::null(),
+            generation: 0,
+            images: Vec::new(),
+            extent,
+            format: vk::SurfaceFormatKHR::default(),
+            needs_rebuild: true,
+
+            old_swapchains: VecDeque::new(),
+        }
+    }
+
+    /// Destroy all swapchain resources. Must not be called while any frames are still in flight on
+    /// the GPU.
+    ///
+    /// # Safety
+    ///
+    /// Access to images obtained from [`images`](Self::images) must be externally synchronized.
+    pub unsafe fn destroy(&mut self, device: &Device, swapchain_fn: &khr::Swapchain) {
+        for frame in &self.frames {
+            device.destroy_fence(frame.complete, None);
+            device.destroy_semaphore(frame.acquire, None);
+        }
+        if self.handle != vk::SwapchainKHR::null() {
+            swapchain_fn.destroy_swapchain(self.handle, None);
+        }
+        for &(swapchain, _) in &self.old_swapchains {
+            swapchain_fn.destroy_swapchain(swapchain, None);
+        }
+    }
+
+    /// Force the swapchain to be rebuilt on the next [`acquire`](Self::acquire) call, passing in
+    /// the surface's current size.
+    pub fn update(&mut self, extent: vk::Extent2D) {
+        self.extent = extent;
+        self.needs_rebuild = true;
+    }
+
+    /// Maximum number of frames that may be concurrently rendered
+    pub fn frames_in_flight(&self) -> usize {
+        self.frames.len()
+    }
+
+    /// Latest set of swapchain images, keyed by [`AcquiredImage::image_index`]
+    pub fn images(&self) -> &[vk::Image] {
+        &self.images
+    }
+
+    /// Format of images in [`images`](Self::images), and the color space that will be used to
+    /// present them
+    pub fn format(&self) -> vk::SurfaceFormatKHR {
+        self.format
+    }
+
+    /// Dimensions of images in [`images`](Self::images)
+    pub fn extent(&self) -> vk::Extent2D {
+        self.extent
+    }
+
+    /// Acquire a swapchain image
+    ///
+    /// Returns [`vk::Result::ERROR_OUT_OF_DATE_KHR`] if and only if the configured format or
+    /// present modes could not be satisfied.
+    ///
+    /// # Safety
+    ///
+    /// `surface_fn` and `swapchain_fn` must have been acquired from `device`, which must match the
+    /// `device` passed to [`new`](Self::new)
+    pub unsafe fn acquire(
+        &mut self,
+        device: &Device,
+        surface_fn: &khr::Surface,
+        swapchain_fn: &khr::Swapchain,
```

Comment:
I think it would better to allow the user to provide a timeout (or perhaps an optional timeout and default to `!0` if not provided).

Reply:
What's the use case? Handling every possible requirement is a non-goal here, since people doing something unusual can always fall back on bare Vulkan.

Reply:
Dropping a frame if one can't be acquired in time or performing other work while waiting. It seems like a pretty simple addition that doesn't complicate this API.

Reply:
Also, the PR is self described as 'lightly opinionated' and not having a timeout feels 'strongly opinionated' (imo).

Reply:
> Dropping a frame if one can't be acquired in time

Ah, fair enough.

> It seems like a pretty simple addition that doesn't complicate this API.

Agreed, given that it provides some value outside of extremely niche cases. Applied.